src/Microsoft.VisualStudio.ProjectSystem.Managed/Tree/ProjectImportsSubTreeProvider.cs

```diff
@@ -0,0 +1,280 @@
+                                else
+                                {
+                                    TearDownTree();
+                                }
+                            }
+                        });
+
+                    return;
+
+                    void SetUpTree()
+                    {
+                        Assumes.Null(_subscription);
+
+                        // Set a visible root
+                        SubmitTreeUpdateAsync(
+                            (currentTree, a, c) =>
+                            {
+                                // Update (make visible) or create a new tree if no prior one exists
+                                IProjectTree tree = currentTree == null
+                                    ? NewTree(Resources.ImportsTreeNodeName, icon: s_rootIcon, flags: ProjectImportsTreeRootFlags)
+                                    : currentTree.Value.Tree.SetVisible(true);
+
+                                return Task.FromResult(new TreeUpdateResult(tree));
+                            });
+
+                        // Subscribe to data to populate the tree and keep it updated with changes
+                        ITargetBlock<IProjectVersionedValue<IProjectImportTreeSnapshot>> actionBlock
+                            = DataflowBlockSlim.CreateActionBlock(
+                                new Action<IProjectVersionedValue<IProjectImportTreeSnapshot>>(SyncTree));
+
+                        IReceivableSourceBlock<IProjectVersionedValue<IProjectImportTreeSnapshot>> importTreeSource
+                            = _projectSubscriptionService.ImportTreeSource.SourceBlock;
+
+                        using (TrySuppressExecutionContextFlow())
+                        {
+                            _subscription = importTreeSource.LinkTo(actionBlock, DataflowOption.PropagateCompletion);
+                        }
+
+                        JoinUpstreamDataSources(_projectSubscriptionService.ImportTreeSource);
+
+                        return;
+
+                        void SyncTree(IProjectVersionedValue<IProjectImportTreeSnapshot> e)
+                        {
+                            SubmitTreeUpdateAsync(
+                                (treeSnapshot, configuredProjectExports, ct) =>
+                                {
+                                    IProjectTree updatedTree = SyncNode(
+                                        imports: e.Value.Value,
+                                        tree: treeSnapshot.Value.Tree);
+
+                                    return Task.FromResult(new TreeUpdateResult(updatedTree, e.DataSourceVersions));
+                                });
+
+                            return;
+
+                            IProjectTree SyncNode(IReadOnlyList<IProjectImportSnapshot> imports, IProjectTree tree)
+                            {
+                                var knownImportPaths = new HashSet<string?>(StringComparers.Paths);
```

Comment:
If imports can be null, then we'll display an empty node in the tree.

Reply:
True. We are setting the path with a non-null value, but the getter is marked as nullable. I've added an `Assumes.NotNullOrEmpty` which satisfies NRT checking.